### ISSUE TITLE
Allow for sound streaming support

### ISSUE THREAD

**Issue by JCash** [defold/defold#9335]
We'd like to support sound streaming of larger sounds.
The goal is to avoid having the entire sound file in-memory, and rather stream portions in and out.

Design:
https://docs.google.com/document/d/1eb6GzHyDZc_L5thgMipKf5_O76eh9FLED55WbhYUSiM/edit?usp=sharing

Sub tasks
* [x] #9855
* [x] #9857
* [x] #9898
* [x] #9897
* [x] #9856
* [x] https://github.com/defold/defold/issues/9958
* [x] https://github.com/defold/defold/issues/9960
* [x] #10010
* [x] #10011

**Comment by Gert-Jan** [comment]
Hi, I'm currently looking into implementing voice chat for a game in Defold. Playing back voice audio data seems not to be possible right now because you can only feed preloaded resources to the sound system. This ticket seems to be a step in a direction where it would be possible to support voice chat. So I guess my questions are these:
1. Is this seriously being pursued? If so when would something like this _roughly_ be released.
2. Can I request a feature on top of this, which is sdk access to the streaming of audio data to a sound component. That way I could stream the voice data and it would be part of the Defold sound mix. There was some talk about it a few years ago, not sure if this went anywhere https://forum.defold.com/t/how-is-audio-exactly-implemented/66655

**Comment by britzl** [comment]
> * Is this seriously being pursued? If so when would something like this _roughly_ be released.

Yes, very much. The work is done in collaboration with a paying partner. I expect it to be available before the end of Q1.

> Can I request a feature on top of this, which is sdk access to the streaming of audio data to a sound component.

I suppose you need this in the dmSDK (ie the C++ API) and not on a Lua level?

**Comment by Gert-Jan** [comment]
> > * Is this seriously being pursued? If so when would something like this _roughly_ be released.
> 
> Yes, very much. The work is done in collaboration with a paying partner. I expect it to be available before the end of Q1.

Great!
> 
> > Can I request a feature on top of this, which is sdk access to the streaming of audio data to a sound component.
> 
> I suppose you need this in the dmSDK (ie the C++ API) and not on a Lua level?

That's how I would use it, yes. A bit like dmGraphics but for sound.

**Comment by britzl** [comment]
> A bit like dmGraphics but for sound.

I can't say exactly what the final API will look like, and if your exact use-case will be covered, but it is likely that there will be an API where you can put chunks of decoded audio into some buffer/sound component and have that played back as a sound component. @JCash is the one in charge of this feature so he may be able to provide more info.

**Comment by Gert-Jan** [comment]
@JCash Just throwing this out there. I have worked on a very similar feature a few years ago, but in C# for XNA. I pretty much used this: https://web.archive.org/web/20201029064413/https://gist.github.com/nickgravelyn/5580531#file-oggsong-cs which uses something like this https://github.com/labnation/MonoGame/blob/master/MonoGame.Framework/Audio/DynamicSoundEffectInstance.cs

In case you'd like some reference :)

**Comment by JCash** [comment]
The initial support for sound streaming are now done.
